Additional password checks

qrainaud opened the pull request:
### New password check option : check if new password contains either name or surname
- Can be enabled/disabled with `$pwd_diff_namesurname` in config
- Backend verification happens in `htdocs/change.php` after submit
- `/rest/v1/getnamesurnamefromlogin.php` api endpoint for future frontend feedback
- Frontend feedback will require ltb-project/ltb-common to be updated, as well as rest api to be enabled

qrainaud committed:
Fix of /rest/v1/include.php : \Ltb\Language::detectLanguage() call as function name was changed to detect_language

qrainaud committed:
Addition of name/surname password check, to make sure neither name nor surname are inside the new password.
- It can be enabled/disabled in config ($pwd_diff_namesurname)
- It requires restapi to be enabled ($use_restapi)
- It uses the api endpoint /rest/v1/checknamesurname.php
- Both english and french languages are currently supported

qrainaud committed:
Added "sameasnamesurname" translations (only french and english supported) for future backend verification (Ltb-common)

qrainaud committed:
Refactor of the checknamesurname api, now returns the name and surname linked to a given login, to be used by the ltb-common library

qrainaud committed:
Addition of backend verification for pwd_diff_namesurname option

qrainaud commented:
For frontend feedback : https://github.com/ltb-project/ltb-common/pull/71

coudot commented:
Hello,

why don't use `$pwd_forbidden_ldap_fields` ?
https://self-service-password.readthedocs.io/en/stable/config_ppolicy.html#forbidden-ldap-fields

qrainaud commented:
Hello,

I didn't notice this. The backend part of my fork is basically useless, but I believe the api as well as the frontend would still be useful as a lot of users tend to put their name or surname in their password, and the feedback message after submission isn't really clear.
And since the message is only shown after submission, they have to rewrite everything.

qrainaud commented:
Hello, let me know what you think.

qrainaud commented:
I can change it so that it checks every forbidden ldap field dynamically when the user is typing its password, and display a feedback similar to "Your password may not contain personal informations". What do you think?

coudot commented:
You can give a try.

I don't swear it will be included as it can be a security flaw: this would allow to discover informations from LDAP directory.

qrainaud commented:
For the security, I will change the api endpoint so that it validates passwords instead of returning LDAP data.
On the api side : to avoid spamming the LDAP directory, I can make a cache system so that it only fetches informations from LDAP when the login changes.

qrainaud committed:
revert backend changes

qrainaud committed:
revert lang, update policyforbiddenldapfields message in french and english languages

qrainaud committed:
revert /rest/v1/getnamesurnamefromlogin.php api endpoint

qrainaud committed:
add /rest/v1/validatepassword.php api endpoint :
- takes a login and a password as x-www-form-urlencoded format parameters
- returns an "isValid" boolean
- uses the project root cache folder to store recent LDAP queries using symfony cache system

The "isValid" boolean returned is :
- true if the login doesn't exist in the LDAP
- true if the login exists and the password doesn't contain any forbidden ldap entry
- false if the login exists and the password contains a forbidden ldap entry

qrainaud committed:
remove unneeded "pwd_diff_namesurname" from changecustompwdfield.php

qrainaud committed:
remove unneeded "pwd_diff_namesurname" from index.php
update "pwd_forbidden_ldap_fields" from pwd_policy_config array in index.php for future template display

qrainaud commented:
Hello again, let me know what you think.

coudot commented:
Sorry for the lack of activity, we are not focused on SSP development for now. I keep this issue open for review
